api/clients/v2/disperser_client.go

```diff
@@ -390,7 +393,137 @@ func (c *disperserClient) GetPaymentState(ctx context.Context) (*disperser_rpc.G
+	allQuorumsReply, err := c.client.GetPaymentStateForAllQuorums(ctx, request)
+	if err != nil {
+		// Check if error is "method not found" or "unimplemented"
+		if isMethodNotFoundError(err) {
+			// Fall back to old method
+			return c.getPaymentStateFromLegacyAPI(ctx, accountID, signature, timestamp)
+		}
+		return nil, err
+	}
+
+	return allQuorumsReply, nil
+}
+
+// this is true if we are targeting a disperser that hasn't upgraded to the new API yet.
+func isMethodNotFoundError(err error) bool {
+	if st, ok := status.FromError(err); ok {
+		return st.Code() == codes.Unimplemented
+	}
+	return false
+}
+
+// getPaymentStateFromLegacyAPI retrieves the payment state from the legacy GetPaymentState grpc method.
+// It is needed until we have upgraded all dispersers (testnet and mainnet) to the new API.
+// Check those endpoints for GetPaymentStateForAllQuorums using:
+// `grpcurl disperser-testnet-holesky.eigenda.xyz:443 list disperser.v2.Disperser`
+// `grpcurl disperser.eigenda.xyz:443 list disperser.v2.Disperser`
+func (c *disperserClient) getPaymentStateFromLegacyAPI(
+	ctx context.Context, accountID gethcommon.Address, signature []byte, timestamp uint64,
+) (*disperser_rpc.GetPaymentStateForAllQuorumsReply, error) {
+	// Convert old request/response format to new format
+	oldRequest := &disperser_rpc.GetPaymentStateRequest{
+		AccountId: accountID.Hex(),
+		Signature: signature,
+		Timestamp: timestamp,
+	}
+
+	oldResult, err := c.client.GetPaymentState(ctx, oldRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// Convert old response to new response format
+	return convertLegacyPaymentStateToNew(oldResult)
+}
+
+// convertLegacyPaymentStateToNew converts the old GetPaymentStateReply to the new GetPaymentStateForAllQuorumsReply format
+func convertLegacyPaymentStateToNew(legacyReply *disperser_rpc.GetPaymentStateReply) (*disperser_rpc.GetPaymentStateForAllQuorumsReply, error) {
+
+	if legacyReply.PaymentGlobalParams == nil {
+		return nil, fmt.Errorf("legacy payment state received from disperser does not contain global params")
+	}
+	// Convert PaymentGlobalParams to PaymentVaultParams
+	var paymentVaultParams *disperser_rpc.PaymentVaultParams
+	{
+		paymentVaultParams = &disperser_rpc.PaymentVaultParams{
+			QuorumPaymentConfigs:  make(map[uint32]*disperser_rpc.PaymentQuorumConfig),
+			QuorumProtocolConfigs: make(map[uint32]*disperser_rpc.PaymentQuorumProtocolConfig),
+			OnDemandQuorumNumbers: legacyReply.PaymentGlobalParams.OnDemandQuorumNumbers,
+		}
+
+		// Apply the global params to all quorums, both on-demand and reservation.
+		onDemandQuorums := legacyReply.PaymentGlobalParams.OnDemandQuorumNumbers
+		if len(onDemandQuorums) == 0 {
+			return nil, fmt.Errorf("no on-demand quorums specified in legacy PaymentGlobalParams received from disperser")
+		}
+		reservationQuorums := legacyReply.Reservation.QuorumNumbers
+		allQuorums := append(reservationQuorums, onDemandQuorums...)
```

Comment:
is appending actually the right operation? Is it guaranteed that there is no overlap between the reservation and the on-demand quorums?

Reply:
[fbbb8ef](https://github.com/Layr-Labs/eigenda/pull/1686/commits/fbbb8ef84f54478b240b8013059877e68fe816b0)